src/operations/Operation.cpp

```diff
@@ -121,6 +121,46 @@ bool Operation::equals(const Operation& op, const Permutation& perm1,
     return false;
   }
 
+  if (isDiagonalGate()) {
+    // check pos. controls and targets together
+    std::set<Qubit> usedQubits1 =
+        perm1.empty()
+            ? getUsedQubits()
+            : getUsedQubits([&perm1](const auto& q) { return perm1.at(q); });
+
+    std::set<Qubit> usedQubits2 =
+        perm2.empty()
+            ? getUsedQubits()
+            : getUsedQubits([&perm2](const auto& q) { return perm2.at(q); });
+
+    if (usedQubits1 != usedQubits2) {
+      return false;
+    }
+
+    std::set<Qubit> negControls1{};
+    for (const auto& control : getControls()) {
+      if (control.type == Control::Type::Neg) {
+        if (perm1.empty()) {
+          negControls1.emplace(control.qubit);
+        } else {
+          negControls1.emplace(perm1.at(control.qubit));
+        }
+      }
+    }
+
+    std::set<Qubit> negControls2{};
+    for (const auto& control : op.getControls()) {
+      if (control.type == Control::Type::Neg) {
+        if (perm2.empty()) {
+          negControls2.emplace(control.qubit);
+        } else {
+          negControls2.emplace(perm2.at(control.qubit));
+        }
+      }
+    }
+
+    return negControls1 == negControls2;
```

Comment:
Sorry for being so pedantic here, but this can be even simpler.

The second block of code can also make use of such a lambda function.
I'd probably define a `perm` function based on whether the permutation is empty. Then you can reuse that function across the entire function here.

Additionally, these simplifications need not be limited to your newly introduced if block. They can be applied across the whole function.

Something like (untested)
```c++
std::function<Qubit(const Qubit)> permFunc1 = perm1.empty()? [](const Qubit q) { return q; } : [&perm1](const Qubit q) { return perm1.at(q); };
```

Reply:
I'll change that in the next commit but then I honestly kind of have enough for only wanting to fix the equivalence check of operations.

Reply:
Fair enough. Sorry for the long detour 😅